wi-fi/lwip/main.c

```diff
@@ -93,94 +138,223 @@ int wifi_ap_set_key(const char *key, size_t len)
+			break;
+		}
+
+		wifi_common.iface.role = CY_LWIP_AP_NW_INTERFACE;
+
+		result = cy_lwip_add_interface(&wifi_common.iface, &ap_addr);
+		if (result != CY_RSLT_SUCCESS) {
+			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't add Wi-Fi interface\n");
+			cybsp_wifi_deinit(wifi_common.iface.whd_iface);
+			cybsp_free();
+			break;
+		}
+
+		result = cy_lwip_network_up(&wifi_common.iface);
+		if (result != CY_RSLT_SUCCESS) {
+			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't bring up Wi-Fi interface\n");
+			cy_lwip_remove_interface(&wifi_common.iface);
+			cybsp_wifi_deinit(wifi_common.iface.whd_iface);
+			cybsp_free();
+			break;
+		}
+
+		result = whd_wifi_init_ap(wifi_common.iface.whd_iface, &ssid, AP_SECURITY_MODE, key, key_len, AP_CHANNEL);
+		if (result != CY_RSLT_SUCCESS) {
+			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't init Wi-Fi AP\n");
+			cy_lwip_network_down(&wifi_common.iface);
+			cy_lwip_remove_interface(&wifi_common.iface);
+			cybsp_wifi_deinit(wifi_common.iface.whd_iface);
+			cybsp_free();
+			break;
+		}
+
+		result = whd_wifi_start_ap(wifi_common.iface.whd_iface);
+		if (result != CY_RSLT_SUCCESS) {
+			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't start Wi-Fi AP\n");
+			cy_lwip_network_down(&wifi_common.iface);
+			cy_lwip_remove_interface(&wifi_common.iface);
+			cybsp_wifi_deinit(wifi_common.iface.whd_iface);
+			cybsp_free();
+			break;
+		}
+
+		started = true;
+	} while (0);
+
+	mutexLock(wifi_common.lock);
+	wifi_common.flags |= started ? WIFI_FLAG_STARTED : WIFI_FLAG_FAILED;
+	condSignal(wifi_common.cond);
+	mutexUnlock(wifi_common.lock);
+
+	if (started) {
+		wifi_ap_main_loop();
+
+		whd_wifi_stop_ap(wifi_common.iface.whd_iface);
 		cy_lwip_network_down(&wifi_common.iface);
 		cy_lwip_remove_interface(&wifi_common.iface);
 		cybsp_wifi_deinit(wifi_common.iface.whd_iface);
 		cybsp_free();
-		return -5;
 	}
 
-	result = whd_wifi_start_ap(wifi_common.iface.whd_iface);
-	if (result != CY_RSLT_SUCCESS) {
-		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't start Wi-Fi AP\n");
-		cy_lwip_network_down(&wifi_common.iface);
-		cy_lwip_remove_interface(&wifi_common.iface);
-		cybsp_wifi_deinit(wifi_common.iface.whd_iface);
-		cybsp_free();
-		return -6;
+	mutexLock(wifi_common.lock);
+	wifi_common.tid = 0;
+	mutexUnlock(wifi_common.lock);
+}
+
+
+static int wifi_ap_start(void)
+{
+	uint8_t flags;
+
+	mutexLock(wifi_common.lock);
+
+	/* reset idle timeout in any case */
+	wifi_common.idle_current = 0;
+
+	if (wifi_common.tid != 0) {
+		mutexUnlock(wifi_common.lock);
+		return 0;
 	}
 
-	wifi_common.running = true;
+	wifi_common.flags = 0;
 
-	return 0;
+	if (sys_thread_opt_new("wifi-ap", wifi_ap_thread, NULL, WIFI_THREAD_STACKSZ, WIFI_THREAD_PRIO, &wifi_common.tid) < 0) {
+		mutexUnlock(wifi_common.lock);
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't create Wi-Fi AP thread\n");
+		return -1;
+	}
+
+	while (wifi_common.flags == 0)
+		condWait(wifi_common.cond, wifi_common.lock, 1000000);
```

Comment:
Do we need timeout if we don't check for timeout condition?
```suggestion
		condWait(wifi_common.cond, wifi_common.lock, 0);
```

Reply:
No, it's remnant of debug code.